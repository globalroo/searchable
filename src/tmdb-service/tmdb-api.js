import {
	getDiscoverMovieEndpoint,
	getMovieByIdEndpoint,
	getMovieCastByIdEndpoint,
	getMovieGenresEndpoint,
	getPersonByIdEndpoint,
	getPersonCreditsByIdEndpoint,
	getPopularMovieEndpoint,
	getRecommendedMoviesByIdEndpoint,
	getSearchMovieEndpoint,
	getSearchMultiEndpoint,
	getTrendingPersonByWeekEndpoint,
	getTrendingMoviesByDay,
	getTrendingMoviesByWeek,
	getTrendingTvByWeekEndpoint,
	getTvByIdEndpoint
} from "./tmdb-config";

export const MEDIA_TYPE = {
	TV: "tv",
	MOVIE: "movie",
	PERSON: "person"
};

export const defaultFailResponse = {
	page: 0,
	results: [],
	total_pages: 0,
	total_results: 0
};

const handleResponse = ({ response, payload }) => ({
	response,
	payload
});

/**
 * Naive safe response with capability to handle basic error state.
 * @async
 * @function safeFetchJson
 * @param {string} url - the endpoint to fetch
 * @param {object} options - fetch request options object default = {},
 * @param {object} failResponse - an expected empty response from tmdb default = defaultFailResponse
 * @return {object} { payload, response } - "payload" - holds JSON from the response or failResponse. "response" - holds the actual response from the server to handle error conditions
 */

export const safeFetchJson = async ({ url, options = {}, failResponse = defaultFailResponse }) => {
	let response, payload;
	try {
		response = await window.fetch(url, options);
		payload = await response.json();
	} catch (error) {
		response = { ok: false, statusText: `${error}` };
		payload = failResponse;
	}
	return handleResponse({ response, payload });
};

// API helpers

export const getPopularMovies = async () =>
	await safeFetchJson({
		url: getPopularMovieEndpoint()
	});

export const getSearchMovieTitleResults = async term =>
	await safeFetchJson({
		url: getSearchMovieEndpoint(term)
	});

export const getSearchMultiResults = async term =>
	await safeFetchJson({
		url: getSearchMultiEndpoint(term)
	});

export const getMovieById = async id =>
	await safeFetchJson({
		url: getMovieByIdEndpoint(id),
		failResponse: {}
	});

export const getRecommendedMoviesById = async id =>
	await safeFetchJson({
		url: getRecommendedMoviesByIdEndpoint(id)
	});

export const getMovieCastById = async id =>
	await safeFetchJson({
		url: getMovieCastByIdEndpoint(id)
	});

export const getPersonById = async id =>
	await safeFetchJson({
		url: getPersonByIdEndpoint(id),
		failResponse: {}
	});

export const getTvById = async id =>
	await safeFetchJson({
		url: getTvByIdEndpoint(id),
		failResponse: {}
	});

export const getDailyTrendingMovies = async () =>
	await safeFetchJson({
		url: getTrendingMoviesByDay()
	});

export const getWeeklyTrendingMovies = async () =>
	await safeFetchJson({
		url: getTrendingMoviesByWeek()
	});

export const getWeeklyTrendingPeople = async () =>
	await safeFetchJson({
		url: getTrendingPersonByWeekEndpoint()
	});

export const getWeeklyTrendingTv = async () =>
	await safeFetchJson({
		url: getTrendingTvByWeekEndpoint()
	});

export const getPersonCreditsById = async id =>
	await safeFetchJson({
		url: getPersonCreditsByIdEndpoint(id)
	});

export const getMovieGenres = async () =>
	await safeFetchJson({
		url: getMovieGenresEndpoint()
	});

export const getDiscoverMoviesByGenreId = async (id, query) => {
	const applyFilter = query
		? query
		: "&sort_by=popularity.desc&include_adult=false&vote_average.gte=6";
	return await safeFetchJson({
		url: `${getDiscoverMovieEndpoint()}${applyFilter}&with_genres=${id}`
	});
};
