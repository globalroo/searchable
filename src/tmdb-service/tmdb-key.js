let API_KEY = ""; //TODO: handle API key

export const getKey = () => {
	return API_KEY;
};
