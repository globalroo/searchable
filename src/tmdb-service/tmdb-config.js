import { getKey } from "./tmdb-key";

export const BASE_URL = "https://api.themoviedb.org/3";
export const BASE_IMG_PATH = "http://image.tmdb.org/t/p";

const API_KEY = getKey();

// Obtain image sizes here
// https://api.themoviedb.org/3/configuration?api_key=

export const API = {
	COMBINED_CREDITS: "/combined_credits",
	CREDITS: "/credits",
	DAY: "/day",
	DISCOVER_MOVIE: "/discover/movie",
	GENRES: "/genre/movie/list",
	MOVIE: "/movie",
	PERSON: "/person",
	POPULAR_MOVIES: "/movie/popular",
	RECOMMENDATIONS: "/recommendations",
	SEARCH_MOVIES: "/search/movie",
	SEARCH_MULTI: "/search/multi",
	SMALL_IMG_WIDTH: "/w300",
	LARGE_IMG_WIDTH: "/w1280",
	TRENDING: "/trending",
	TV: "/tv",
	WEEK: "/week"
};

export const getAuthorisedEndpoint = endpoint => `${BASE_URL}${endpoint}?api_key=${API_KEY}`;

export const getSmallPosterImage = imagePath =>
	`${BASE_IMG_PATH}${API.SMALL_IMG_WIDTH}/${imagePath}`;

export const getLargeImage = imagePath => `${BASE_IMG_PATH}${API.LARGE_IMG_WIDTH}/${imagePath}`;

export const getMovieByIdEndpoint = id =>
	getAuthorisedEndpoint(`${API.MOVIE}/${id}`) + "&append_to_response=recommendations,credits";

export const getMovieCastByIdEndpoint = id =>
	getAuthorisedEndpoint(`${API.MOVIE}/${id}${API.CREDITS}`);

export const getPopularMovieEndpoint = () => getAuthorisedEndpoint(API.POPULAR_MOVIES);

export const getRecommendedMoviesByIdEndpoint = id =>
	getAuthorisedEndpoint(`${API.MOVIE}/${id}${API.RECOMMENDATIONS}`);

export const getSearchMovieEndpoint = searchTerm =>
	`${getAuthorisedEndpoint(API.SEARCH_MOVIES)}&query=${searchTerm}`;

export const getSearchMultiEndpoint = searchTerm =>
	`${getAuthorisedEndpoint(API.SEARCH_MULTI)}&query=${searchTerm}`;

export const getTrendingMoviesByDay = () =>
	getAuthorisedEndpoint(`${API.TRENDING}${API.MOVIE}${API.DAY}`);

export const getTrendingMoviesByWeek = () =>
	getAuthorisedEndpoint(`${API.TRENDING}${API.MOVIE}${API.WEEK}`);

export const getTrendingPersonByWeekEndpoint = () =>
	getAuthorisedEndpoint(`${API.TRENDING}${API.PERSON}${API.WEEK}`);

export const getTrendingTvByWeekEndpoint = () =>
	getAuthorisedEndpoint(`${API.TRENDING}${API.TV}${API.WEEK}`);

export const getPersonByIdEndpoint = id =>
	getAuthorisedEndpoint(`${API.PERSON}/${id}`) + "&append_to_response=images";

export const getPersonCreditsByIdEndpoint = id =>
	getAuthorisedEndpoint(`${API.PERSON}/${id}${API.COMBINED_CREDITS}`);

export const getTvByIdEndpoint = id =>
	getAuthorisedEndpoint(`${API.TV}/${id}`) + "&append_to_response=recommendations,credits";

export const getMovieGenresEndpoint = () => getAuthorisedEndpoint(`${API.GENRES}`);
export const getDiscoverMovieEndpoint = () => getAuthorisedEndpoint(`${API.DISCOVER_MOVIE}`);
