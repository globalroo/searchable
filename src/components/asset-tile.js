import React from "react";

import { getSmallPosterImage } from "src/tmdb-service/tmdb-config";
import { useImageLoader } from "src/hooks/image-loader";
import { navigateTo } from "src/helpers/navigate";
import { GridListTile, GridListTileBar } from "@material-ui/core";

const styles = {
	gridListTile: {
		height: "225px",
		width: "150px"
	},
	spacer: { padding: "5px" }
};

export const getTitle = asset => (asset.title ? asset.title : asset.name);
export const getImageSrc = asset => (asset.poster_path ? asset.poster_path : asset.profile_path);

export const AssetTile = ({ asset }) => {
	const imageSrc = getImageSrc(asset);
	const [image] = useImageLoader(getSmallPosterImage(imageSrc));
	const title = getTitle(asset);
	return (
		<GridListTile
			style={{ ...styles.spacer, ...styles.gridListTile }}
			onClick={() => navigateTo(asset)}
			className="pointer-hover image-tile"
			component="div"
		>
			<img src={image} alt={title} />
			<GridListTileBar title={title} />
		</GridListTile>
	);
};
