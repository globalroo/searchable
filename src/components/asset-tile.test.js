import React from "react";
import { render, cleanup, fireEvent } from "react-testing-library";

import { AssetTile } from "./asset-tile";
import mockMovie from "src/test-helpers/mock-movie";
import { MEDIA_TYPE } from "src/tmdb-service/tmdb-api";

const mockNavigateSpy = jest.fn();

jest.mock("src/helpers/navigate", () => ({
	navigateTo: val => {
		mockNavigateSpy(val);
	}
}));

describe("Asset Tile tests", () => {
	afterEach(cleanup);
	it("Renders an AssetTile component", () => {
		const asset = { ...mockMovie, media_type: "Defined" };
		const { getByTestId } = render(<AssetTile asset={asset} />);
		const assetTile = getByTestId(`${asset.id}_${asset.title}`);
		expect(assetTile).toMatchSnapshot();
	});
	it("Renders an AssetScroller component with a non-default media type", () => {
		const asset = {
			...mockMovie,
			title: undefined,
			name: "person_name",
			poster_path: undefined,
			profile_path: "Person type tile"
		};
		const { getByTestId } = render(<AssetTile asset={asset} media_type={"TEST"} />);
		const assetTile = getByTestId(`${asset.id}_${asset.name}`);
		expect(assetTile).toMatchSnapshot();
	});
	it("Expect component to navigate when clicked", () => {
		const asset = { ...mockMovie, media_type: MEDIA_TYPE.MOVIE };
		const { getByTestId } = render(<AssetTile asset={asset} />);
		const assetTile = getByTestId(`${asset.id}_${asset.title}`);
		fireEvent.click(assetTile, { button: 1 });
		expect(mockNavigateSpy).toHaveBeenCalledWith(asset);
	});
});
