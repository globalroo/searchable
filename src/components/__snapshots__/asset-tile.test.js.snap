// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Asset Tile tests Renders an AssetScroller component with a non-default media type 1`] = `
<div
  class="MuiGridListTile-root-16 pointer-hover image-tile"
  data-testid="226486_person_name"
  style="padding: 5px; height: 225px; width: 150px;"
>
  <div
    class="MuiGridListTile-tile-17"
  >
    <img
      alt="person_name"
      src="http://image.tmdb.org/t/p/w300/Person type tile"
    />
    <div
      class="MuiGridListTileBar-root-20 MuiGridListTileBar-titlePositionBottom-21"
    >
      <div
        class="MuiGridListTileBar-titleWrap-24"
      >
        <div
          class="MuiGridListTileBar-title-27"
        >
          person_name
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Asset Tile tests Renders an AssetTile component 1`] = `
<div
  class="MuiGridListTile-root-1 pointer-hover image-tile"
  data-testid="226486_Tammy"
  style="padding: 5px; height: 225px; width: 150px;"
>
  <div
    class="MuiGridListTile-tile-2"
  >
    <img
      alt="Tammy"
      src="http://image.tmdb.org/t/p/w300//cq6wvDqETqJXgpQplkL0FBw2leM.jpg"
    />
    <div
      class="MuiGridListTileBar-root-5 MuiGridListTileBar-titlePositionBottom-6"
    >
      <div
        class="MuiGridListTileBar-titleWrap-9"
      >
        <div
          class="MuiGridListTileBar-title-12"
        >
          Tammy
        </div>
      </div>
    </div>
  </div>
</div>
`;
